### ISSUE TITLE
KBucket splitting is flawed.

### ISSUE THREAD

**Issue by StuartFarmer** [bmuller/kademlia#73]
It seems that when you split a KBucket, it uses an unbounded value provided by the node id. Splitting buckets uses the node id to determine which bucket to put things into. However, due to the way it is implemented, this does not occur. Consider this:

```
n1 = Node(b'ag')
n2 = Node(b'gggggggg')
```
The binary representation of these nodes, which is used when splitting, is the following
```
b'ag' -> '1100000010111'
b'gggggggg'' -> '11000000010110010101010001001101100010010100101011111'
```
Therefore, these pieces of data have a depth of 12. This is extremely close together in terms of the protocol. However, the pieces of data are in reality massively different and should be 100% stored in different buckets.

The protocol needs to enforce either left padding or hashes to standardize the size of the node id to maintain correctness.

**Comment by StuartFarmer** [comment]
Furthermore, it could be argued that data is never put into the bucket that is outside of the bucket's range. This is what the RoutingTable object expects (https://github.com/bmuller/kademlia/blob/master/kademlia/routing.py#L177).

However, there is no enforcement that the node is within range when it is being added to the bucket, so there is no guarantee that this is true at the time of splitting (https://github.com/bmuller/kademlia/blob/master/kademlia/routing.py#L53).

**Comment by bmuller** [comment]
Thanks @StuartFarmer - I'm having trouble following your first comment.  In the first sentence, for instance - the k-bucket's upper range will always be bounded by the max value of a node's id, which is the max value of a SHA1 digest (160 bits), so I'm not sure how it could ever be "unbounded."  

The first bucket therefore has bounds of 0 to 2^160.  When the bucket is split, it is split in half at the midway point (the first split would produce two buckets, one with ids between 0 and 2^159, and the other with ids between 2^159 and 2^160).

Also, one item that I think might be conflated is the shared bit prefix and the distance measure used (XOR).  The shared prefix (or depth) is used as a part of the bucket splitting, but isn't the same thing as the distance measure used between nodes (XOR).

Maybe a good place to start is to see whether you ultimately think that the paper is flawed, or if you think this implementation fails to follow the paper.  If the former, then any changes would be [outside of the scope of this project](https://github.com/bmuller/kademlia#fidelity-to-original-paper).  If the latter, then it would be super helpful to clarify where this process diverges from what's described in the paper (most helpful would be pointing to the section/paragraph/etc).  The bucket splitting and distance measure should both be what's described in the paper's details - though I definitely may have missed something or implemented incorrectly.

RE the second comment - the router will [only add a node to a bucket whose bounds contain that node's id](https://github.com/bmuller/kademlia/blob/master/kademlia/routing.py#L157:L161).  This is a guarantee.

**Comment by StuartFarmer** [comment]
What I'm saying is that if you're using this on the front end, then maybe `digest` is in fact called, but if you are doing lower level integrations, which is what I am doing, then you can see that KBuckets can have nodes added to them without `digest` being called and the issue rises.

For example: https://github.com/bmuller/kademlia/blob/master/kademlia/node.py#L6

You can set the node ID to any byte string. There are no restrictions on length. This is the core issue that I'm presenting.

**Comment by bmuller** [comment]
@StuartFarmer what's the use case that would require instantiating `Node`s directly?  In normal use of this lib, that class should never be used directly.

**Comment by StuartFarmer** [comment]
We aren't using any of the networking code, just using the core Kademlia data structures. Therefore, none of the 'front-end' security checks are going to occur in our situation. We're going through QA testing by writing unit tests on our entire codebase and came across this during that time.

The best implementation of this algorithm would not require the data structures to be used in a particular way. A potential simple solution would be to run `digest` in the `__init__` and call it a day.

**Comment by bmuller** [comment]
@StuartFarmer many of the places where `Node`s are instantiated only have access to the 160-bit ID (and not the pre-digested value before being SHA1'ed) - so I'm not sure how to handle those situations  if the change you suggest were to be made.

Additionally - I see in your [cilantro](https://github.com/Lamden/cilantro) project you've copied a number of files from this project.  I'm happy that you've been able to find a use for the code - and the inclusion is permitted under the license - but [with the one condition that you include the copyright notice and permission notice](https://github.com/bmuller/kademlia/blob/master/LICENSE#L11) in those copies.  I (and the other many contributors to this project) would greatly appreciate the inclusion the copyright notice and permission notice somewhere in the copied code.  Thanks, and again happy to see usage!

**Comment by StuartFarmer** [comment]
We'll include this notice :)

Regardless of how you do it, enforcing the bit length of the identifier is going to solve these issues. People who are bypassing the front-end security measures are prone to the errors I've described and the protocol will break.

**Comment by bmuller** [comment]
Fixed in f218b83.  Added docs to make it clear what's expected in the constructor for anyone who wants to use that class directly at such a low level.
